### ISSUE TITLE
Depth of application

### ISSUE THREAD

**Issue by lionelgit** [hadley/lowliner#22]
In case of complicated list structures, it could be nice to be able to specify the depth of application of `map()` and friends.

I've been trying this:

```
recurse <- function(.x) {
  call <- sys.call(-1)
  call <- match.call(match.fun(call[[1]]), call)
  call$.x <- as.name(".x")
  call$.depth <- call$.depth - 1

  lapply(.x, function(.x, cl) eval(cl), call)
}
```

Then, at the start of each mapping function, add a call to recurse
```
map <- function(.x, .f, ..., .depth = 1) {
  if (.depth > 1) {
    return(recurse(.x))
  }

  .f <- as_function(.f)
  lapply(.x, .f, ...) %>% output_hook(.x)
}
```

So far I tried it with `map()` and `keep()` and this works well.

```
deep_list <- list(
  a = list(
    aa = list(1:2, 3:4),
    ab = list(5:6, 7:8)
  ),
  b = list(
    ba = list(9:10, 11:12),
    bb = list(13:14, 15:16)
  )
)

deep_list %>%
  map(sum, .depth = 3) %>%
  map(splat(mean), .depth = 2)
```

If you're interested in this functionality, I can test it with other lowliner functions in proper unit tests and submit a PR.

**Comment by lionelgit** [comment]
Another possibility would be to use a factory adverb to adjust any function that takes a list.

```{r}
at_depth <- function(.depth, .f) {
  function(...) {
    dots <- list(...)
    x <- dots[[1]]
    args <- dots[-1]

    recurse <- function(x, depth) {
      if (depth > 1) {
        lapply(x, recurse, depth = depth - 1)
      } else {
        x <- list(x)
        names(x) <- names2(dots)[1]
        do.call(.f, c(x, args))
      }
    }

    recurse(x, .depth)
  }
}
```

Instead of
```{r}
deep_list <- list(
  a = list(
    aa = list(1:2, 3:4),
    ab = list(5:6, 7:8)
  ),
  b = list(
    ba = list(9:10, 11:12),
    bb = list(13:14, 15:16)
  )
)

deep_list %>%
  map(sum, .depth = 3) %>%
  map(splat(mean), .depth = 2)
```
we have
```{r}
deep_list %>%
  at_depth(3, map)(sum) %>%
  at_depth(2, map)(splat(mean))
```

**Comment by hadley** [comment]
I would really really really prefer to avoid NSE in lowliner (or at least minimise at as much as possible).  And I find the idea of `at_depth` hard to grasp. You can currently accomplish the same thing by nesting `map()`, e.g.

```R
deep_list %>%
  map(. %>% map(. %>% map_v(sum))) %>% 
  map(. %>% map(mean))
```

That's not super easy to read, but it does give you control of which sort of map is used at each level, and I don't think we should reward the use of highly complicated data structures.

**Comment by lionelgit** [comment]
hmm except in my first attempt to have a reusable `recurse()`, there is no NSE involved in either the `.depth` arg or the adverb `at_depth()` I think. They are regular instances of functional programming no?

I see your point about complicated data structures, but even at depth 2 or 3, `at_depth()` is easier to read than nested maps. Anyway, the cool thing with `at_depth()` is that it can be used for any functions that takes a list, so it does not need to be included in lowliner to be used easily.

**Comment by hadley** [comment]
Oh sorry I missed that implementation, that does look rather good. I'd be happy to review a pull request

Although you need to carefully consider the use of `do.call()` - http://rpubs.com/hadley/do-call2.

**Comment by lionelgit** [comment]
Interesting. So instead of
```
do.call(.f, c(x, args))
```
I should have written
```
args <- c(x, args)
do.call(".f", quote(args))
```

**Comment by lionelgit** [comment]
Closed by #33
